### PR TITLE
Docker fix

### PR DESCRIPTION
**Pull request by blakeflei**
This is really about creating a sharable, working docker implementation of reports that runs reliably without any changes to the underlying codebase. It also provides a straightforward, reliable means for testing locally before we make future changes. The existing implementation via requirements.txt was non-functional on M1 macs, and there are updates to the readme to facilitate collaboration. There will also be communications about which service account should be used moving forward - if you are unsure please ask.

Docker compose is still used, and configured so debugging can be accomplished in jupyter notebook via forwarded ports. The command to start a docker compose in the container terminal is included with the readme and mentioned below.

#### Running
This was tested locally, but reports generation should be easily replicated locally via the following (this info is also in updated readme):
1. Install docker locally, set resources as desired (i.e. enable 6 cores if you have an 8 core machine, etc.)
2. Configure the service account json, place in `$HOME/.config/gcloud/service-account-file.json`
3. Clone repo and navigate to root folder in a terminal:

```
# Create docker image+container and start the container, this drops the terminal into the container:
docker-compose run --service-ports calitp_reports /bin/bash

# Remove all prior reports, if any
make clean 

# For a single month, edit the non-commented first line in the Make file so the month is accurately reflected:
# For example, for March 2022, change the line:
# NOTEBOOKS=$(subst parameters.json,index.html,$(wildcard outputs/2021/09/*/parameters.json))
# to:
# NOTEBOOKS=$(subst parameters.json,index.html,$(wildcard outputs/2022/03/*/parameters.json))
make all -j 8
``` 
For rerunning all historical months, there is a script `run_all_months.py` that can be used via:
```
python run_all_months.py
```

#### Debugging 
For debugging, an interactive jupyter notebook server within the container can be started via:
```
jupyter notebook --ip 0.0.0.0 --port 8891
```
The generated URL can be copied and pasted into the host browser.

#### Additional PR Info
Fixes [data-infra 1362](https://github.com/cal-itp/data-infra/issues/1362)

Type of change: Bug fix

### CODE REVIEW

**Review by themightychris — APPROVED** [review]

**Comment by lauriemerrell** [comment]
I am fine to merge this as-is, but just one note -- `make clean` will empty `outputs/`. If you then run `make sync`, you will sync your (possibly incomplete) `outputs/` to GCS (test bucket) and overwrite any existing data. For the test data this is fine but for `make sync-prod` this is very dangerous. Not sure if it's worth adding some note to the README within this PR about using extreme caution with `make sync`, `make sync-prod`, etc.

**Comment by blakeflei** [comment]
> I am fine to merge this as-is, but just one note -- `make clean` will empty `outputs/`. If you then run `make sync`, you will sync your (possibly incomplete) `outputs/` to GCS (test bucket) and overwrite any existing data. For the test data this is fine but for `make sync-prod` this is very dangerous. Not sure if it's worth adding some note to the README within this PR about using extreme caution with `make sync`, `make sync-prod`, etc.

The GCP sdk is not installed in the docker container and any gsutil rsync commands will currently fail; using docker is actually overly cautious in this aspect and we need the sdk in place to do the syncs. I'm very open to adding the sdk install to the docker image but left it off because of that the reason you mention.

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    volumes:
+      - $HOME/.config/gcloud:/home/reports/.config/gcloud
+      - $PWD/reports:/home/reports/app
+    ports:
+      # To run container jupyter notebooks on the host
+      - 8891:8891
+    environment:
+      - GOOGLE_APPLICATION_CREDENTIALS=/home/reports/.config/gcloud/service-account-file.json
+      # Set the service key path for calitp-py
+      - CALITP_SERVICE_KEY_PATH=/home/reports/.config/gcloud/service-account-file.json
```
Comment (acouch) [inline docker-compose.yml:18]: Notice these are the same values, should `GOOGLE_APPLICATION_CREDENTIALS` be the `application_default_credentials.json` file?

Comment (blakeflei) [reply docker-compose.yml:18]: They point to the same thing as a service account. That variable isn't needed though, and can be removed.

**Comment by acouch** [comment]
>  make sync-prod this is very dangerous

**Comment by Nkdiaz** [comment]
I am able to run docker locally but I am running into the same problem on docker as I am when running the virtual environment locally. When running make generate_parameters, I get the following errors:
ModuleNotFoundError: No module named 'pyodbc'
make: *** [Makefile:20: generate_parameters] Error 1
Please see ticket https://github.com/cal-itp/reports/issues/117

**Comment by acouch** [comment]
Are there instructions for downloading the `service-account-file.json` ?  I get access denied when I go to the services page in GCP. I might not have the correct instructions.

Could we parameterize the Makefile so you can pass a month and year `make all -j 8 MONTH=12 YEAR=2022` and an ENTRYPOINT to the docker file so we could run the entire process by passing args to the container?

**Review by Nkdiaz — APPROVED** [review]
I was able to run it locally on my machine

**Comment by lauriemerrell** [comment]
Just want to second @acouch's question above -- it sounds from Slack like he ended up passing his GCloud credentials another way. I think it would be good to clarify the recommended approach for authenticating before merging this PR.

**Comment by blakeflei** [comment]
> Are there instructions for downloading the `service-account-file.json` ? I get access denied when I go to the services page in GCP. I might not have the correct instructions.
> 
> Could we parameterize the Makefile so you can pass a month and year `make all -j 8 MONTH=12 YEAR=2022` and an ENTRYPOINT to the docker file so we could run the entire process by passing args to the container?

Thank you for pointing out the credentials issue; I originally created a separate service account but user account credentials are preferred so I've updated the docker-compose and Readme instructions.

I totally agree the convenience fix of a default current-month-only run would be great - I'd like to fix a few things in the Makefile, but those edits are out of scope for this PR. This is focused on just getting docker to run reliably.

**Review by lauriemerrell — APPROVED** [review]

**Comment by acouch** [comment]
I have been able to get it working with the latest commit. 

I got what seems like a timeout error generating the report:

```
   await ensure_async(self.kc.wait_for_ready(timeout=self.startup_timeout))
  File "/home/reports/venv/lib/python3.8/site-packages/nbclient/util.py", line 89, in ensure_async
    result = await obj
  File "/home/reports/venv/lib/python3.8/site-packages/jupyter_client/client.py", line 180, in _async_wait_for_ready
    raise RuntimeError("Kernel died before replying to kernel_info")
RuntimeError: Kernel died before replying to kernel_info
```

This was not repeated rerunning the same month. Addressing that could be an optimization for a future PR.

I would like to parameterize the makefile but that could wait for another PR.

**Review by acouch — APPROVED** [review]

**Pull request merged**